klog.go

```diff
@@ -520,6 +532,75 @@ func (l *loggingT) setVState(verbosity Level, filter []modulePat, setFilter bool
 
 var timeNow = time.Now // Stubbed out for testing.
 
+// CaptureState gathers information about all current klog settings and
+// returns a function that, when called, restores those settings.
+// This does not invalidate caches. The restore function may get called
+// more than once.
+func CaptureState() func() {
+	logging.mu.Lock()
+	defer logging.mu.Unlock()
+	flushDRunning := logging.flushD.isRunning()
+	moduleSpec := logging.vmodule.serialize()
+	logger := globalLogger
+	loggerOptions := globalLoggerOptions
+	maxSize := MaxSize
+	state := loggingT{
```

Comment:
Hiding `loggingT` behind a function seems a nice way to avoid breaking changes, however I would want to make sure we should reconsider how the testing should look like.

Looking at https://github.com/kubernetes/kubernetes/blob/b1e130fe83156783153538b6d79821c2fdaa85bb/staging/src/k8s.io/component-base/logs/json/klog_test.go#L276-L285 current approach is:

```
fs := flag.FlagSet{}
klog.InitFlags(&fs)
fs.Set("v", "1")

defer func() {
	err := fs.Set("v", "0")
	if err != nil {
		t.Fatalf("Failed to reset verbosity to 0")
	}
}()
```

With `CaptureState` I expect it to change to:

```
fs := flag.FlagSet{}
klog.InitFlags(&fs)
defer klog.CaptureState()()
fs.Set("v", "1")
```
So the main way to interface with klog configuration will be still via `flag.FlagSet`. Could we maybe think about separating runtime struct `loggingT` from configuration like in was done in k/k?

Reply:
That sounds like a major API change for klog. I don't think we should do that no matter how much nicer a new configuration API would be, simply because klog is so widely used and because adding "experimental" APIs is almost impossible outside of test code.